rotkehlchen/exchanges/ftx.py

```diff
@@ -0,0 +1,477 @@
+        """
+        request_verb = "GET"
+
+        request_url = '/api/' + endpoint
+        options = {}
+        if limit is not None:
+            options['limit'] = limit
+        if start_time is not None:
+            options['start_time'] = start_time
+        if end_time is not None:
+            options['end_time'] = end_time
+
+        if len(options.keys()):
+            request_url += '?' + urlencode(options)
+
+        timestamp = int(time.time() * 1000)
+        signature_payload = f'{timestamp}{request_verb}{request_url}'.encode()
+
+        signature = hmac.new(self.secret, signature_payload, 'sha256').hexdigest()
+        log.debug('FTX API query', request_url=request_url)
+
+        self.session.headers.update({
+            'FTX-KEY': self.api_key,
+            'FTX-SIGN': signature,
+            'FTX-TS': str(timestamp),
+        })
+
+        full_url = self.base_uri + request_url
+        try:
+            response = self.session.get(full_url)
+        except requests.exceptions.RequestException as e:
+            raise RemoteError(f'FTX API request failed due to {str(e)}') from e
+
+        if response.status_code == 429:
+            raise RemoteError('FTX rate limit exceeded')
+
+        if response.status_code == 401:
+            raise FTXLoginError('Authentication failed.')
+
+        if response.status_code != 200:
+            raise RemoteError(
+                f'FTX query {full_url} responded with error status code: '
+                f'{response.status_code} and text: {response.text}',
+            )
+
+        try:
+            json_ret = jsonloads_dict(response.text)
+        except JSONDecodeError as e:
+            raise RemoteError(f'FTX returned invalid JSON response: {response.text}') from e
+
+        if 'result' not in json_ret:
+            raise RemoteError(f'FTX json response does not contain data: {response.text}')
+
+        return json_ret['result']
+
+    def _api_query(
+            self,
+            endpoint: str,
+            start_time: int = None,
+            end_time: int = None,
+            limit: int = 100,
+            paginate: bool = True,
+    ) -> Any:
+        if not paginate:
+            return self._make_request(
+                endpoint=endpoint,
+                limit=limit,
+                start_time=start_time,
+                end_time=end_time,
+            )
+
+        # If there is pagination
+        new_end_time = end_time
+        ids = set()
+        final_data = []
+        while True:
+            step = self._make_request(
+                endpoint=endpoint,
+                limit=limit,
+                start_time=start_time,
+                end_time=new_end_time,
+            )
+
+            if not isinstance(step, list):
+                self.msg_aggregator.add_error('Trying pagination in an invalid endpoint.')
+                break
+
+            # remove possible duplicates
```

Comment:
I see you must have gotten this from the example [here](https://github.com/ftexchange/ftx/blob/master/rest/client.py#L163) but I am not sure ... why would there be duplicates? I guess if their example has it, it's probably a good idea to follow but I don't understand why.

Do you?

Can you also add a link to the example's code here? Otherwise this code makes no sense.

Reply:
Sure. From my experiments I didn't get duplicates but since their own code has this check I wanted keep it there

Reply:
Duplicates happen in the limits of the page. Since the limit is inclusive you get one duplicate entry each time

Reply:
> Duplicates happen in the limits of the page. Since the limit is inclusive you get one duplicate entry each time

If this is indeed always the case can't you then use a different limit, or take `data[:-1]` or `data[1:]` depending on which side the pagination moves to?

Sounds like a better approach than keeping an extra mapping in a set, no?

Reply:
I believe so but I'm not sure if this approach is interesting to be defensive against possible duplicates from the API if this can happen

Reply:
Anyway let's leave it like that for now then. I don't want to overthink it.